modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java

```diff
@@ -296,45 +277,72 @@ private void handleDeleteAllCommand(DeleteAllCommand cmd, long commandIndex) {
-                    .forEachRemaining(
-                            lock -> {
-                                if (lock.lockMode() == LockMode.EXCLUSIVE && lock.lockKey().key() instanceof RowId) {
-                                    storage.commitWrite((RowId) lock.lockKey().key(), new Timestamp(txId));
-                                }
-                            }
-                    );
-        } else /*if (txManager.state(txId) == TxState.ABORTED)*/ {
-            lockManager.locks(txId)
-                    .forEachRemaining(
-                            lock -> {
-                                if (lock.lockMode() == LockMode.EXCLUSIVE && lock.lockKey().key() instanceof RowId) {
-                                    storage.abortWrite((RowId) lock.lockKey().key());
-                                }
-                            }
-                    );
+    private void handleFinishTxCommand(FinishTxCommand cmd, long commandIndex) throws IgniteInternalException {
+        UUID txId = cmd.txId();
+
+        TxState stateToSet = cmd.commit() ? TxState.COMMITED : TxState.ABORTED;
+
+        boolean txStateChangeRes = txStateStorage.compareAndSet(
+                txId,
+                null,
+                new TxMeta(
+                        stateToSet,
+                        cmd.replicationGroupIds(),
+                        cmd.commitTimestamp()
+                )
+        );
+
+        if (!txStateChangeRes) {
+            UUID traceId = UUID.randomUUID();
+
+            String errorMsg = format("Fail to finish the transaction txId = {} because of inconsistent state = {},"
+                    + " expected state = null, state to set = {}", txId, txStateStorage.get(txId), stateToSet);
+
+            IgniteInternalException stateChangeException = new IgniteInternalException(traceId, TX_UNEXPECTED_STATE, errorMsg);
+
+            // Exception is explicitly logged because otherwise it can be lost if it did not occur on the leader.
+            LOG.error(errorMsg);
+
+            throw stateChangeException;
+        }
+
+        storage.lastAppliedIndex(commandIndex);
+    }
+
+
+    /**
+     * Handler for the {@link TxCleanupCommand}.
+     *
+     * @param cmd Command.
+     * @param commandIndex Index of the RAFT command.
+     */
+    private void handleTxCleanupCommand(TxCleanupCommand cmd, long commandIndex) {
+        UUID txId = cmd.txId();
+
+        Set<RowId> removedRowIds = txsRemovedKeys.get(txId);
+
+        Set<RowId> insertedRowIds = txsInsertedKeys.get(txId);
+
+        Set<RowId> pendingRowIds = new HashSet<>(removedRowIds);
+        pendingRowIds.addAll(insertedRowIds);
+
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-17577 Use HybridTimestamp instead.
+        Timestamp commitTimestamp = new Timestamp(cmd.commitTimestamp().getPhysical(), cmd.commitTimestamp().getLogical());
+
+        if (cmd.commit()) {
+            pendingRowIds.forEach(rowId -> storage.commitWrite(rowId, commitTimestamp));
```

Comment:
Need to invoke commitWrite/abortWrite for updated rows, not only for inserted and removed.

Reply:
Thanks, fixed.